ADC: replace ioctl with uorb message

SalimTerryLi opened the pull request:
This is spitted from #13833 .

This PR focus on replace ioctls with uORB message. Some existing modules/drivers are affected.

SalimTerryLi commented:
Device ID is set but not used. All logic kept as before. This should be safe to merge.

SalimTerryLi commented:
There are two things to be checked:
1. whether [old ioctl message](https://github.com/PX4/Firmware/blob/86eff44a5efa01273788bc46e0001c4102cafff2/src/drivers/drv_adc.h#L53) manages channel_id field correctly? I mean that `adc_msg[A].am_channel == A` should be true if that channel is valid.
2. Do all boards have the same 3v3 ADC vref? If not, there should be some macro defined cases in ADC.cpp.

bkueng commented on src/modules/battery_status/analog_battery.cpp:
```diff
@@ -22,10 +22,6 @@ AnalogBattery::AnalogBattery(int index, ModuleParams *parent) :
 
-	_analog_param_handles.cnt_v_volt = param_find("BAT_CNT_V_VOLT");
-
-	_analog_param_handles.cnt_v_curr = param_find("BAT_CNT_V_CURR");
```
Can you go through the airframe configs (`posix-configs/` and `ROMFS/px4fmu_common/init.d/airframes`), and remove the parameters and change the voltage dividers accordingly to get to the same result?

SalimTerryLi replied:
It will be the same result if ADC driver provides correct v_ref. I'm not sure how those annoying macros is formed. It should work, but we need test each device.

SalimTerryLi replied:
"ADC count to voltage" actually is LSB of ADC.

SalimTerryLi replied:
Instead we should change `BAT_V_DIV` and `BAT_A_PER_V`

SalimTerryLi replied:
Or we need to fix the correct v_ref reported by `drivers/adc`

SalimTerryLi replied:
I suggest that adding extra macros which define their ADC v_ref in each `board_config.h`, or creating a new low level API in `drv_adc.h`, or developing new ADC drivers for each device :( .
There is no much boards depend on `BAT_CNT_V_*`.

bkueng commented on src/modules/battery_status/analog_battery.cpp:
```diff
@@ -34,21 +30,25 @@ AnalogBattery::AnalogBattery(int index, ModuleParams *parent) :
+	snprintf(param_name, sizeof(param_name), "BAT%d_V_CHANNEL", index);
+	_analog_param_handles.v_channel = param_find(param_name);
+
+	snprintf(param_name, sizeof(param_name), "BAT%d_I_CHANNEL", index);
```
I'd rather not add this as long as we don't need it. Even `BAT%d_V_CHANNEL` might not be that useful.

SalimTerryLi replied:
Maybe I'll need it in the future..... To add support for current sensing........ Can we leave it there as it has no negative effect?

bkueng replied:
There is: the more params there are the more difficult it is to setup a system.

SalimTerryLi replied:
These two parameters can be ignored by default. Change when necessary. I have exposed extra ADC channels to connect current sensors for that current monitor can not be integrated into my board.

bkueng replied:
Are you saying you need `BAT%d_V_CHANNEL`? None of the airframes/boards currently has `BAT%d_ADC_CHANNEL` set.
If you know you'll need them, let's keep them, otherwise we can drop them.

SalimTerryLi replied:
V_CHANNEL is required by some boards, for sure. I'm asking to add I_CHANNEL, which is required by me...

SalimTerryLi replied:
bbblue requires `BAT_ADC_CHANNEL`, which is changed to `BATi_V_CHANNEL`

bkueng commented on src/drivers/rc_input/RCInput.cpp:
```diff
@@ -308,21 +308,20 @@ void RCInput::Run()
 
 		if (_adc_sub.update(&adc)) {
-			const unsigned adc_chans = sizeof(adc.channel_id) / sizeof(adc.channel_id[0]);
+			if (adc.channel_id[ADC_RC_RSSI_CHANNEL] == ADC_RC_RSSI_CHANNEL) {
```
This is not necessarily true. You need to find the channel.

SalimTerryLi replied:
Does this means that array index is not the same as channel ID?

bkueng replied:
Correct, that is why there was a loop around this before.

bkueng commented on src/modules/battery_status/battery_status.cpp:
```diff
@@ -171,11 +170,7 @@ BatteryStatus::parameter_update_poll(bool forced)
-
-	/* read all channels available */
-	int ret = px4_read(_adc_fd, &buf_adc, sizeof(buf_adc));
+	adc_report_s adc_report = {};
```
```suggestion
	adc_report_s adc_report;
```
And please move this down before `if (_adc_report_sub.update(&adc_report)) {`

bkueng commented on msg/adc_report.msg:
```diff
@@ -1,3 +1,6 @@
+uint32 device_id		# unique device ID for the sensor that does not change between power cycles
+int16[12] channel_id		# ADC channel IDs, negative for non-existent, should be kept same as array index
+int32[12] raw_data		# ADC channel raw value, accept negative value, valid if channel ID is positive
+uint32[12] resolution		# ADC channel resolution, valid if channel ID is positive
```
resolution is the same for all channels, and can be uint16

SalimTerryLi replied:
But [here](https://github.com/PX4/Firmware/blob/a3710fcdefa2eb109f5250908c7b2ebf0301c9fc/src/drivers/drv_adc.h#L93)?

bkueng replied:
You can change it there as well, or keep everything uint32, but we don't need that much.

SalimTerryLi replied:
An ADC with 16bit precision will have this value of 65536, which is overflowing the uint16_t.

bkueng replied:
I thought we weren't at the limit yet. Alright then it's clear.

SalimTerryLi replied:
> 
> 
> I thought we weren't at the limit yet. Alright then it's clear.

What? Isn't stm32h7 already using 16bit ADC? https://github.com/PX4/Firmware/blob/0d36e5094b4655c960846d481f62361d40cf1d9f/platforms/nuttx/src/px4/stm/stm32h7/adc/adc.cpp#L312-L315

bkueng replied:
It is, so we go with uint32.

bkueng commented on src/modules/battery_status/battery_status.cpp:
```diff
@@ -194,63 +189,69 @@ BatteryStatus::adc_poll()
-						/* Voltage in volts */
-						bat_voltage_adc_readings[b] = buf_adc[i].am_data;
+			/* look for specific channels and process the raw voltage to measurement data */
+			if (adc_report.channel_id[_analogBatteries[b]->get_voltage_channel()] == _analogBatteries[b]->get_voltage_channel()) {
```
`_analogBatteries[b]->get_voltage_channel()` is the ADC channel, which is not necessarily the index of `channel_id`.

bkueng commented on src/modules/battery_status/battery_status.cpp:
```diff
@@ -194,63 +189,69 @@ BatteryStatus::adc_poll()
 		}
 	}
+
+	for (int b = 0; b < BOARD_NUMBER_BRICKS; b++) {
```
Keep this within `if (_adc_report_sub.update(&adc_report)) {`

SalimTerryLi commented on src/drivers/adc/ADC.cpp:
```diff
@@ -148,7 +149,10 @@ void ADC::update_adc_report(hrt_abstime now)
 		adc.channel_id[i] = _samples[i].am_channel;
-		adc.channel_value[i] = _samples[i].am_data * 3.3f / px4_arch_adc_dn_fullcount();
+		adc.raw_data[i] = _samples[i].am_data;
+		adc.v_ref[i] = 3.3f;    // TODO: Is this correct? All boards have 3v3 v_ref?
```
I go through the code base but only find [this](https://github.com/PX4/Firmware/blob/ca0b5700ab0ba6dd2c088564dc13988af27d3263/platforms/common/include/px4_platform_common/board_common.h#L150-L156) reference. The value could only be 3v3.

SalimTerryLi commented:
[This](https://github.com/PX4/Firmware/blob/8df22541ef4b4e334a14ad900e1d70601620c0e1/ROMFS/px4fmu_common/init.d/airframes/4016_holybro_px4vision#L132-L133) is the last reference to that parameter. But I cannot find anything in the code which provides ADC specification such as sample precision and voltage reference. 0.0008 mostly indicate that the board is using 3v3 as vref with 2^12 bits precision but I cannot find anything related in code.

SalimTerryLi commented:
> Good rework, thanks.
> What is the testing you have done?

Currently I have no hardware to test. This pr is diverged from a previous one so my test code on my platform could not be uploaded. If needed, I could commit them temporarily and then revert them for a clean PR.

bkueng commented:
> This is the last reference to that parameter. But I cannot find anything in the code which provides ADC specification such as sample precision and voltage reference. 0.0008 mostly indicate that the board is using 3v3 as vref with 2^12 bits precision but I cannot find anything related in code.
@bkueng Sorry but I'm unfamiliar with these devices. Any help is welcome.

They should not have been added to that board, you can just remove them from there (the default should be used).

SalimTerryLi commented:
> 
> 
> > This is the last reference to that parameter. But I cannot find anything in the code which provides ADC specification such as sample precision and voltage reference. 0.0008 mostly indicate that the board is using 3v3 as vref with 2^12 bits precision but I cannot find anything related in code.
> > @bkueng Sorry but I'm unfamiliar with these devices. Any help is welcome.
> 
> They should not have been added to that board, you can just remove them from there (the default should be used).

Acknowledged. I'm contacting dagar to get ADC reference voltage properly provided for each board.

davids5 commented on src/drivers/adc/ADC.cpp:
```diff
@@ -148,9 +149,28 @@ void ADC::update_adc_report(hrt_abstime now)
-		adc.channel_value[i] = _samples[i].am_data * 3.3f / px4_arch_adc_dn_fullcount();
+		adc.raw_data[i] = _samples[i].am_data;
+		adc.v_ref[i] = 3.3f;	// TODO: fill the true v_ref, maybe new macros for embed device?
+#ifdef ADC_BATTERY_VOLTAGE_CHANNEL
```
@SalimTerryLi - why is this needed? We want to remove the BOARD_ADC_POS_REF_V_FOR_VOLTAGE_CHAN and BOARD_ADC_POS_REF_V_FOR_CURRENT_CHAN  and take the ref from the ADC via the new API

SalimTerryLi replied:
The new PR has not been merged yet. I'm waiting for that to provide such a API....

davids5 replied:
@SalimTerryLi - it is merged now

SalimTerryLi replied:
Done with 86808ed

davids5 commented on src/drivers/rc_input/RCInput.cpp:
```diff
@@ -308,16 +308,17 @@ void RCInput::Run()
 
 					if (_analog_rc_rssi_volt < 0.0f) {
-						_analog_rc_rssi_volt = adc.channel_value[i];
+						_analog_rc_rssi_volt = adc_volt;
```
This looks like it repeated the last sample.  Now it does not

SalimTerryLi replied:
I cannot find those lines of code. Instead they already are.
https://github.com/SalimTerryLi/Firmware/blob/86808ed736efb40f1f832869de786ffedce3ed27/src/drivers/rc_input/RCInput.cpp#L317-L319
Not sure what is happening here.

davids5 replied:
Here is the old code https://github.com/PX4/Firmware/blob/bcfc0e2f5625cc975ecbd304e9cfeaac51d362bf/src/drivers/rc_input/RCInput.cpp#L310-L326

It looks like it has been changed, and now does not act the same. @dagar ?

dagar replied:
I'm not sure what the question/confusion is. Here's the corresponding chunk of logic. https://github.com/PX4/Firmware/blob/4c1adc088fa8e7cc682e16582300cbf91cd84f67/src/drivers/rc_input/RCInput.cpp#L200-L214

It might be easier and simpler to do the adc_report copy and minimal work directly in `RCInput::fill_rc_in` when rssi hasn't been provided.

SalimTerryLi replied:
> I'm not sure what the question/confusion is. Here's the corresponding chunk of logic.
> 
> https://github.com/PX4/Firmware/blob/4c1adc088fa8e7cc682e16582300cbf91cd84f67/src/drivers/rc_input/RCInput.cpp#L200-L214
> 
> It might be easier and simpler to do the adc_report copy and minimal work directly in `RCInput::fill_rc_in` when rssi hasn't been provided.

hard-coded base voltage which is used to calculate rssi. This may work, may not either.
But in this PR I only do the replacement of data source so this might not something to be discussed here?

SalimTerryLi replied:
@davids5 Please forgive me, I'm really confused with what you said. In the `RCInput.cpp`, I only replace old ioctl call with new uORB message, no functional logic changed. In other word, this PR is not aimed to improve anything about RSSI, but only keep its existing function works with new uORB subscription. No doubt `adc.channel_value[i]` is equal to `adc_volt`.

SalimTerryLi commented:
Solved conflict with master https://github.com/PX4/Firmware/commit/4c1adc088fa8e7cc682e16582300cbf91cd84f67

SalimTerryLi commented:
It's ready. @dagar 
If someone could provide a test on the real hardware to prove its correctness would be better. Sorry for that I do not have any device to do such a test. If analog battery is monitored correctly, then this is safe to merge.

SalimTerryLi commented:
It's ready. @dagar 
If someone could provide a test on the real hardware to prove its correctness would be better. Sorry for that I do not have any device to do such a test. If analog battery is monitored correctly, then this is safe to merge.

davids5 commented on src/drivers/adc/ADC.cpp:
```diff
@@ -148,9 +149,13 @@ void ADC::update_adc_report(hrt_abstime now)
 		adc.channel_id[i] = _samples[i].am_channel;
-		adc.channel_value[i] = _samples[i].am_data * 3.3f / px4_arch_adc_dn_fullcount();
+		adc.raw_data[i] = _samples[i].am_data;
+		//adc.channel_value[i] = _samples[i].am_data * 3.3f / px4_arch_adc_dn_fullcount();
```
Should we remove this?

SalimTerryLi replied:
Done.

SalimTerryLi commented:
Tested with new [ADS1115 driver](https://github.com/SalimTerryLi/Firmware/tree/pr-ads1115_driver-new/src/drivers/adc/ADS1115) which is not included in this PR. Battery voltage is displayed correctly in QGC.

jinger26 commented:
@dagar @davids5   shall we ping @PX4/testflights  to test this PR?

davids5 commented on src/modules/battery_status/battery_status.cpp:
```diff
@@ -190,33 +183,43 @@ BatteryStatus::adc_poll()
+	adc_report_s adc_report;
+
+	if (_adc_report_sub.update(&adc_report)) {
+
```
This change look incorrect.

There is 1 set of channels. Of which both bricks values are listed in that set of channels.
It was once through the the list of channels and selected the one brick's values.
Now it it doing multiple passes on the channels

SalimTerryLi replied:
Fixed with 63c08cb

SalimTerryLi commented on src/drivers/adc/ADC.cpp:
```diff
@@ -139,6 +139,7 @@ void ADC::update_adc_report(hrt_abstime now)
 {
 	adc_report_s adc = {};
 	adc.timestamp = now;
+	adc.device_id = BUILTIN_ADC_DEVID;
```
So, which value should be used here now? I have no idea...

davids5 replied:
This can be left for the next PR. This needs to be an API from the adc driver.

davids5 commented on src/modules/battery_status/battery_status.cpp:
```diff
@@ -190,34 +183,40 @@ BatteryStatus::adc_poll()
-					 * supply on power controller) that is valid.
+		/* Read add channels we got */
+		for (int i = 0; i < PX4_MAX_ADC_CHANNELS; ++i) {
+			for (int b = 0; b < BOARD_NUMBER_BRICKS; b++) {
```
The '{' was scoping, and it was removed.

SalimTerryLi replied:
Oh sorry I missed it

davids5 replied:
No. It was in original code but not deleted

dannyfpv commented:
tested on pixhawk4 v5 f-450


Flight Card 1

Modes Tested:
Position Mode: Good.
Altitude Mode: Good.
Stabilized Mode: Good.

Procedure:
Arm and Take off in position mode, after flying for approximately one minute, switched to altitude then stabilized mode.

Notes:
No issues were noted, good flight in general.

Logs:
https://review.px4.io/plot_app?log=3766bbd8-6451-440e-a07c-a18a1fdb536a


Flight Card 2

Modes Tested
Mission Plan Mode (Automated): Good.
RTL (Return To Land): Good.

Procedure

Armed form QGC.
Took-off on mission plan mode
. Note that commands were sent form QGC (Armed and takeoff)
Once all waypoint completed vehicle switched to RTL and landed as expected.
Good flight in general.
Note:
No issues were noted, good flight in general.

Logs:
https://review.px4.io/plot_app?log=ad1757a8-a2a6-408a-abc2-334363a30781



Flight Card 3
Modes Tested
Position Mode: Good.
Mission Plan Mode (Automated): Good.
RTL (Return To Land): Good.

Procedure
Arm and Take off in position mode, after flying for approximately one minute, switched to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, 
Good flight in general.

Note:
No issues were noted, good flight in general.

Logs:
https://review.px4.io/plot_app?log=627d653a-a54d-46d4-9a70-48c75ef7b72e


Flight Card 4
Modes Tested
Fail safe:The radio control turned off and the vehicle returned and landed correctly.
Tested good.

Logs:
https://review.px4.io/plot_app?log=14674a30-b19d-45d4-b0c5-88a9d47504ee


master daily log:
https://review.px4.io/plot_app?log=614301d0-ce86-4ecf-9ddb-a37e48736a97

jorge789 commented:
### Tested on PixRacer V4

**Flight Card 1**

**Modes Tested:**
Position Mode: Good.
Altitude Mode: 
Stabilized Mode: Good.

**Procedure:**
Arm and Take off in position mode, after flying for approximately one minute, switched to altitude then stabilized mode.

**Notes:**
**In altitude mode, the vehicle did not respect the height that continued to increase in pitch forward**

**Logs:** https://review.px4.io/plot_app?log=6d8ae576-1188-4f0b-958e-3dd0399cb85f

**Flight Card 2**

**Modes Tested**
Mission Plan Mode (Automated): Good.
RTL (Return To Land): Good.

**Procedure**
Armed form QGC.
Took-off on mission plan mode
. Note that commands were sent form QGC (Armed and takeoff)
Once all waypoint completed vehicle switched to RTL and landed as expected.
**Note:**
No issues were noted, good flight in general.

**Logs:** https://review.px4.io/plot_app?log=50170288-2b0d-420b-a30d-a1f55cc59f22


**Flight Card 3**

**Modes Tested**
Position Mode: Good.
Mission Plan Mode (Automated): Good.
RTL (Return To Land): Good.

**Procedure**
Arm and Take off in position mode, after flying for approximately one minute, switched to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC,once completed all waypoint then swith RTL.
**Note:**
No issues were noted, good flight in general.

**Logs:** https://review.px4.io/plot_app?log=b97752b6-d3a9-40e2-915e-64a8bb3e4e0b

**Flight Card 4**

**- Modes Tested**
Position Mode: Good.
Failsafe RTL (Return To Land): Not returning after RC shutdown.
**- Procedure**
Arm and take off in position mode, after flying approximately a few seconds, turned off the RC to activate the fail-safe RTL, wait for the vehicle to return and land.
**- Notes**
No issue noted, good flight in general.

**Logs:** https://review.px4.io/plot_app?log=def9cbfc-fa88-4e7b-9829-9ac9a2943e59

**- Modes Tested**
Position Mode: Good.
Altitude Mode: 
Stabilized Mode: Good.
Mission Plan Mode (Automated): Good.
RTL (Return To Land): Good.
**- Procedure**
Arm and Take off in position mode, after flying for approximately one minute, switched to altitude then stabilized mode proceed to switch to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC,  once completed all waypoint then swith RTL..
**- Notes**
**In altitude mode, the vehicle did not respect the height that continued to increase in pitch forward**

### Master comparison
**Log** https://review.px4.io/plot_app?log=a839c371-2ca5-407a-a394-e6326f22d218

### Tested on CUAV nano V5

**Flight Card 1**

**Modes Tested:**
Position Mode: Good.
Altitude Mode: Good.
Stabilized Mode: Good.

**Procedure:**
Arm and Take off in position mode, after flying for approximately one minute, switched to altitude then stabilized mode.

**Notes:**
No issues were noted, good flight in general.

**Logs:** https://review.px4.io/plot_app?log=24b5a036-8f0f-4986-9cae-6839783f9ae3

**Flight Card 2**

**Modes Tested**
Mission Plan Mode (Automated): Good.
RTL (Return To Land): Good.

**Procedure**
Armed form QGC.
Took-off on mission plan mode
. Note that commands were sent form QGC (Armed and takeoff)
Once all waypoint completed vehicle switched to RTL and landed as expected.
**Note:**
No issues were noted, good flight in general.

**Logs:**  https://review.px4.io/plot_app?log=d59a43a0-7788-4bbe-8f36-ad75dabb16d3

**Flight Card 3**

**Modes Tested**
Position Mode: Good.
Mission Plan Mode (Automated): Good.
RTL (Return To Land): Good.

**Procedure**
Arm and Take off in position mode, after flying for approximately one minute, switched to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC,once completed all waypoint then swith RTL.
**Note:**
No issues were noted, good flight in general.

**Logs:**  https://review.px4.io/plot_app?log=cb3821ce-bbf5-479e-93f7-e559b5795bef

**Flight Card 4**

**- Modes Tested**
Position Mode: Good.
Failsafe RTL (Return To Land): Not returning after RC shutdown.
**- Procedure**
Arm and take off in position mode, after flying approximately a few seconds, turned off the RC to activate the fail-safe RTL, wait for the vehicle to return and land.
**- Notes**
No issue noted, good flight in general.

**Logs:**  https://review.px4.io/plot_app?log=686c9864-bf47-45bf-ba5b-61dad6dee54b

**- Modes Tested**
Position Mode: Good.
Altitude Mode: Good.
Stabilized Mode: Good.
Mission Plan Mode (Automated): Good.
RTL (Return To Land): Good.
**- Procedure**
Arm and Take off in position mode, after flying for approximately one minute, switched to altitude then stabilized mode proceed to switch to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC,  once completed all waypoint then swith RTL..
**- Notes**
No issues noted, good flight in general.

### Master comparison
**Log** https://review.px4.io/plot_app?log=e93a68b0-a64b-4c38-9e8a-6da079c2a7f4

dannyfpv commented:
tested on pixhawk 4 mini v5 f500
**Modes Tested**

Position Mode: Good.
Altitude Mode: Good.
Stabilized Mode: Good.
Mission Plan Mode (Automated): Good.
RTL: Good.

**- Procedure**
Arm and Take off in position mode, after flying for approximately one minute, switched to altitude then stabilized mode proceed to switch to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoin activate RTL.

**Notes:**
No issues noted, good flight in general.

**Log:**

https://review.px4.io/plot_app?log=2ae25c0f-a07e-402e-8248-ada5332d0974

master daily log:
https://review.px4.io/plot_app?log=082f3f34-b203-4d7a-be9a-caeb816ab7a3

Junkim3DR commented:
### Tested on NXP FMUK66 v3

Vehicle did not arm pop up error shown below.

[image](https://user-images.githubusercontent.com/47370446/75825896-2efd9800-5d5b-11ea-859e-88c1919f30bf.png)

Proceeded to test on master and no issue noted.

**Logs**
- PR14087: https://review.px4.io/plot_app?log=fc825089-da5b-41bb-815f-d930c3213205
- Master: https://review.px4.io/plot_app?log=a0950d90-d5b5-40cc-b439-90717bfe0821

### Tested on Pixhawk 3 v4Pro

**Flight Card 1**
 - Modes Tested
Position Mode: Good.
Altitude Mode: Good.
Stabilized Mode: Good.

- Procedure
Arm and Take off in position mode, after flying for approximately one minute, switched to altitude then stabilized mode.

- Notes
No issues noted, good flight in general.

- Log
https://review.px4.io/plot_app?log=ae55bea8-5eb9-42e1-be29-af5bdcee7b78

**Flight Card 2**
- Modes Tested
Mission Plan Mode (Automated): Good.
RTL (Return To Land): Good.

- Procedure
Arm and Take off in mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once all waypoints are completed make sure vehicle triggers RTL.

- Notes
Good flight in general.

- Log
https://review.px4.io/plot_app?log=9d6cc0b1-f9b0-4cd3-948d-77e30c127801


**Flight Card 3**
- Modes Tested
Position Mode: Good.
Mission Plan Mode (Automated): Good.
RTL (Return To Land): Good.

- ProcedureArm and Take off in position mode, after flying for approximately one minute, switched to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC.

- Notes
Good flight in general.

-Log
https://review.px4.io/plot_app?log=ed59a773-bca2-46da-99b9-3f700a3f3da8

SalimTerryLi commented:
> ### Tested on NXP FMUK66 v3
> 
> Vehicle did not arm pop up error shown below.
> 
> [image](https://user-images.githubusercontent.com/47370446/75825896-2efd9800-5d5b-11ea-859e-88c1919f30bf.png)
> 
> Proceeded to test on master and no issue noted.
> 
> **Logs**
> 
>     * PR14087: https://review.px4.io/plot_app?log=fc825089-da5b-41bb-815f-d930c3213205
> 
>     * Master: https://review.px4.io/plot_app?log=a0950d90-d5b5-40cc-b439-90717bfe0821
> 
> 
> ### Tested on Pixhawk 3 v4Pro
> 
> **Flight Card 1**
> 
>     * Modes Tested
>       Position Mode: Good.
>       Altitude Mode: Good.
>       Stabilized Mode: Good.
> 
>     * Procedure
>       Arm and Take off in position mode, after flying for approximately one minute, switched to altitude then stabilized mode.
> 
>     * Notes
>       No issues noted, good flight in general.
> 
>     * Log
>       https://review.px4.io/plot_app?log=ae55bea8-5eb9-42e1-be29-af5bdcee7b78
> 
> 
> **Flight Card 2**
> 
>     * Modes Tested
>       Mission Plan Mode (Automated): Good.
>       RTL (Return To Land): Good.
> 
>     * Procedure
>       Arm and Take off in mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once all waypoints are completed make sure vehicle triggers RTL.
> 
>     * Notes
>       Good flight in general.
> 
>     * Log
>       https://review.px4.io/plot_app?log=9d6cc0b1-f9b0-4cd3-948d-77e30c127801
> 
> 
> **Flight Card 3**
> 
>     * Modes Tested
>       Position Mode: Good.
>       Mission Plan Mode (Automated): Good.
>       RTL (Return To Land): Good.
> 
>     * ProcedureArm and Take off in position mode, after flying for approximately one minute, switched to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC.
> 
>     * Notes
>       Good flight in general.
> 
> 
> -Log
> https://review.px4.io/plot_app?log=ed59a773-bca2-46da-99b9-3f700a3f3da8

On the k66 I found it strange that param "BAT_A_PER_V" stores NaN. I can not reproduce that on my device as it will always be 0 default. I'm not sure whether is this problem that blocking battery status or not.

mrpollo commented:
Discussed on the dev call (03/04/20), @davids5 will perform bench testing to figure out what the problem is with k66

davids5 commented:
@SalimTerryLi  I tested this just now on a K66

Master:

```
listener battery_status

TOPIC: battery_status
 battery_status_s
        timestamp: 13779788  (0.007279 seconds ago)
        voltage_v: 15.6210
        voltage_filtered_v: 15.5830
        current_a: 0.0124
        current_filtered_a: 0.0124
        average_current_a: 0.0000
        discharged_mah: 0.0393
        remaining: 1.0000
        scale: 1.0000
        temperature: nan
        cell_count: 0
        voltage_cell_v: [0.0000, 0.0000, 0.0000, 0.0000]
        max_cell_voltage_delta: 0.0000
        capacity: 0
        cycle_count: 0
        run_time_to_empty: 0
        average_time_to_empty: 0
        serial_number: 0
        connected: True
        system_source: True
        priority: 0
        id: 1
        is_powering_off: False
        warning: 0
 adc test
0: 1701  10: 1901  11: 1  12: 3060  13: 2716  16: 2083  18: 230  23: 1  26: 836
0: 1701  10: 1901  11: 1  12: 3057  13: 2713  16: 2085  18: 230  23: 1  26: 836
0: 1701  10: 1901  11: 1  12: 3056  13: 2716  16: 2083  18: 229  23: 1  26: 836
```

This Branch:
```
listener battery_status

TOPIC: battery_status
 battery_status_s
        timestamp: 3746999  (0.005231 seconds ago)
        voltage_v: 0.0000
        voltage_filtered_v: 0.0000
        current_a: -1.0000
        current_filtered_a: 0.0000
        average_current_a: 0.0000
        discharged_mah: 0.0000
        remaining: 1.0000
        scale: 1.0000
        temperature: nan
        cell_count: 0
        voltage_cell_v: [0.0000, 0.0000, 0.0000, 0.0000]
        max_cell_voltage_delta: 0.0000
        capacity: 0
        cycle_count: 0
        run_time_to_empty: 0
        average_time_to_empty: 0
        serial_number: 0
        connected: False
        system_source: False
        priority: 0
        id: 1
        is_powering_off: False
        warning: 0

adc test
0: 1693  10: 1927  11: 1  12: 3069  13: 2623  16: 1975  18: 265  23: 4  26: 844
0: 1693  10: 1931  11: 1  12: 3066  13: 2629  16: 1967  18: 267  23: 3  26: 844
0: 1693  10: 1929  11: 1  12: 3068  13: 2621  16: 1969  18: 265  23: 3  26: 843
0: 1693  10: 1927  11: 1  12: 3066  13: 2621  16: 1967  18: 265  23: 3  26: 843
0: 1693  10: 1927  11: 1  12: 3068  13: 2629  16: 1967  18: 265  23: 4  26: 843
```

To interpret the ADC values on the K66
```
#define ADC_BATTERY_VOLTAGE_CHANNEL ADC1_CH(10)     /* BAT_VSENS        85   PTB4  ADC1_SE10 */
#define ADC_BATTERY_CURRENT_CHANNEL ADC1_CH(11)     /* BAT_ISENS        86   PTB5  ADC1_SE11 */
```

Any Ideas?

davids5 commented on src/modules/battery_status/battery_status.cpp:
```diff
@@ -190,34 +183,40 @@ BatteryStatus::adc_poll()
+
+				if (adc_report.channel_id[i] == _analogBatteries[b]->get_voltage_channel()) {
+					/* Voltage in volts */
+					bat_voltage_adc_readings[b] = adc_report.raw_data[_analogBatteries[b]->get_voltage_channel()] *
```
```suggestion
					bat_voltage_adc_readings[b] = adc_report.raw_data[i] *
```

davids5 commented on src/modules/battery_status/battery_status.cpp:
```diff
@@ -190,34 +183,40 @@ BatteryStatus::adc_poll()
+								      adc_report.resolution;
+
+				} else if (adc_report.channel_id[i] == _analogBatteries[b]->get_current_channel()) {
+					bat_current_adc_readings[b] = adc_report.raw_data[_analogBatteries[b]->get_current_channel()] *
```
```suggestion
					bat_current_adc_readings[b] = adc_report.raw_data[i] *
```

davids5 replied:
@SalimTerryLi This is what caused the error on the K66. But the param migration is also not correct. Please have a look at it.

davids5 commented:
> Discussed on the dev call (03/04/20), @davids5 will perform bench testing to figure out what the problem is with k66

@mrpollo - Done. @PX4/testflights - please retest on FMUv5 with 2 batteries and a K66.

@dagar please have a look at the migration question. At a minimum I would think migrateParam should be renamed to migrateParamAndGet or this stuff we be deleted and param will not be initialized .

dannyfpv commented:
Tested on Pixhawk4 v5 f-450
Indoor Flight Test
Modes Tested
Stabilized Mode: Good.

Procedure
Arm and Take off in stabilized mode.

Notes
No issues noted, good flight in general.

Log
https://review.px4.io/plot_app?log=da8e0d6b-aaf2-488b-acc2-87baf538a6a3

Junkim3DR commented:
### Testedon NXP FMUK66 v3

**Indoor test flight.**
**Mode tested**
- Stabilized

**Procedure**
- Takeoff and landing in Stabilized

**Note**
- No issue noted.

**Log**
- https://review.px4.io/plot_app?log=3743a97b-a985-4107-b688-15a153fdd5c0

SalimTerryLi commented:
Those two logs indicate that this PR now perform the same as master. Are we ready?

bkueng commented on src/drivers/rc_input/RCInput.cpp:
```diff
@@ -308,16 +308,17 @@ void RCInput::Run()
-			for (unsigned i = 0; i < adc_chans; i++) {
+			for (int i = 0; i < PX4_MAX_ADC_CHANNELS; ++i) {
 				if (adc.channel_id[i] == ADC_RC_RSSI_CHANNEL) {
+					float adc_volt = adc.raw_data[ADC_RC_RSSI_CHANNEL] *
```
```suggestion
					float adc_volt = adc.raw_data[i] *
```

SalimTerryLi commented on src/drivers/rc_input/RCInput.cpp:
```diff
@@ -308,16 +308,17 @@ void RCInput::Run()
-			const unsigned adc_chans = sizeof(adc.channel_id) / sizeof(adc.channel_id[0]);
-
-			for (unsigned i = 0; i < adc_chans; i++) {
+			for (int i = 0; i < PX4_MAX_ADC_CHANNELS; ++i) {
```
`PX4_MAX_ADC_CHANNELS` comes from `drv_adc.h`, which indicates the max capacity of `adc_report` topic, They are not the same value in the most case?

bkueng replied:
They are now, but I want to have the guarantee that they really stay the same (if one is changed, the other will be as well).

SalimTerryLi replied:
That means we should use definition from `adc_report.msg` as primary declaration and deprecate `PX4_MAX_ADC_CHANNELS`?

SalimTerryLi replied:
I get to understand your idea. Thanks for point out that.

bkueng replied:
Yes maybe we can just remove `PX4_MAX_ADC_CHANNELS`. I have not checked in detail where it's used everywhere.

SalimTerryLi replied:
32c19a4

SalimTerryLi commented:
@bkueng V_CHANNEL is required by some boards, for sure. I'm asking to add I_CHANNEL, which is required by me...

davids5 commented:
@dagar - Please review the last 2 test logs FMUv5 and K66 - K66 logs Current, V5 does not. Is that also the case on master, and why?

SalimTerryLi commented:
https://github.com/PX4/Firmware/blob/2dccc0dcfc49da59e7fd718531dd9deed007157c/src/modules/sensors/sensors.cpp#L363-L414
still needs some works to make it read from uORB.

SalimTerryLi commented:
Ready to merge (again).
Each review would come up with new issue and this PR already has been improved a lot.
By now, I'm sure battery status works fine and have the same behavior as master branch. RC RSSI isn't tested but its logic is simple enough to assume its correctness. Analog differential pressure sensor as well. I also modified two tests under `systemcmds/tests`. If them are fine there will be no problem.

davids5 commented on src/drivers/drv_adc.h:
```diff
@@ -42,14 +42,15 @@
  *
  */
-#define PX4_MAX_ADC_CHANNELS 12
+#define PX4_MAX_ADC_CHANNELS ((int)(sizeof(adc_report_s::channel_id)/sizeof(*adc_report_s::channel_id)))
```
This looks wrong:
(sizeof(adc_report_s::channel_id)/sizeof(*adc_report_s::channel_id[0])))

There is an ArraySize Macro in the tree px4_macros.h. IMHO I think code using ArraySize(adc_report_s::channel_id)) is better then having the #define (That define is needed in the ADC driver as a constant.  - then to insure the the ADC driver has enough channels a static assert should be used as @bkueng suggested.

SalimTerryLi replied:
Thanks for advice. I'm using `arraySize()` instead of that. But is it necessary to apply a static assert? I don't agree with that, because the max number of available channel is only defined in `adc_report.msg` now.

SalimTerryLi commented:
Any further suggestion?

bkueng commented on src/modules/battery_status/analog_battery.cpp:
```diff
@@ -34,21 +28,25 @@ AnalogBattery::AnalogBattery(int index, ModuleParams *parent) :
+	float voltage_v = voltage_raw * _analog_params.v_div;
+	float current_a = (current_raw - _analog_params.v_offs_cur) * _analog_params.a_per_v;
 
+	// TODO: voltage_v should be changed to voltage_raw?
```
Can we get this resolved? @davids5 ?

SalimTerryLi replied:
`voltage_v` is the voltage that v_div applied. But the condition here seems refer to the original ADC sampled voltage.

davids5 replied:
I need more context to answer this, I am lost in the naming.  What is a raw voltage? Is this unfiltered? What is v_div? 

This comparison should be voltage to voltage unfiltered.

SalimTerryLi replied:
> I need more context to answer this, I am lost in the naming. What is a raw voltage? Is this unfiltered? What is v_div?
> 
> This comparison should be voltage to voltage unfiltered.

"raw voltage" refers to the value provided by ADC driver directly. If ADC has 3v3 vref, raw voltage would within that range. "v_div" is the short form of "voltage divider", which is the extra resistances that divides battery voltage level down to ADC range. Both of them are unfiltered.

davids5 replied:
Thank you! @SalimTerryLi . So let's delete the comment. In the HW we referrer to the divided voltage as V5_SCALED. This is the effect of the voltage divider. Should we consider renaming the v_div to something line v5_scale_factor?

bkueng commented on src/modules/sensors/sensors.cpp:
```diff
@@ -148,6 +148,7 @@ class Sensors : public ModuleBase<Sensors>, public ModuleParams, public px4::Sch
 
```
Can you remove `_adc_fd` above?

Then it should also be possible to remove `ADC0_DEVICE_PATH` completely.

SalimTerryLi replied:
Done.

bkueng commented on src/systemcmds/tests/test_adc.cpp:
```diff
@@ -37,56 +37,36 @@
-		if (count < 0) {
-			goto errout_with_dev;
-		}
+	px4_usleep(1000000);	// sleep 1s and wait for adc report
```
We shouldn't wait that long, it just delays tests.

SalimTerryLi replied:
Done.
